File: golang/recordio_test.go
package recordio

import (
	"fmt"
	"testing"
)

func TestRecordIO(t *testing.T) {
	r1 := &Document{
		Docid:   int64(10),
		Payload: "hello 10",
	}
	r1.Links = new(Links)
	r1.Links.Backward = make([]int64, 0, 10)
	r1.Links.Forward = make([]int64, 0, 10)
	r1.Links.Backward = append(r1.Links.Backward, 10)
	r1.Links.Backward = append(r1.Links.Backward, 13)
	r1.Links.Forward = append(r1.Links.Forward, 15)
	r1.Links.Forward = append(r1.Links.Forward, 17)

	url := "http://www.weibo.com/zhihu?from=feed&loc=nickname"
	r2 := &Document{
		Docid:   int64(20),
		Payload: "hello 20",
	}
	r2.Name = make([]*Name, 0, 10)
	name := new(Name)
	name.Url = url
	r2.Name = append(r2.Name, name)

	r3 := &Document{
		Docid:   int64(30),
		Payload: "hello 30",
	}

	datapath := "/tmp/recordiotest.dat"
	writer, err := NewRecordWriter(datapath)
	if err != nil {
		t.Fatal(err)
	}

	if _, err := writer.WriteProtocolMessage(r1); err != nil {
		t.Fatal(err)
	}
	if _, err := writer.WriteProtocolMessage(r2); err != nil {
		t.Fatal(err)
	}
	if _, err := writer.WriteProtocolMessage(r3); err != nil {
		t.Fatal(err)
	}
	writer.Finish() // close underlying fd

	msg := &Document{}
	reader, err := NewRecordReader(datapath)
	if err != nil {
		t.Fatal(err)
	}

	// Read Record r1
	if _, err := reader.ReadProtocolMessage(msg); err != nil {
		t.Fatal(err)
	}
	if msg.Docid != r1.Docid || msg.Payload != r1.Payload {
		t.Fatalf("data mismatch: %v != %v", msg, r1)
	}
	fmt.Println(msg.Links.Backward) // [10 13]
	fmt.Println(msg.Links.Forward)  // [15 17]

	// Read Record r2
	if _, err := reader.ReadProtocolMessage(msg); err != nil {
		t.Fatal(err)
	}
	if msg.Docid != r2.Docid || msg.Payload != r2.Payload {
		t.Fatalf("data mismatch: %v != %v", msg, r2)
	}
	fmt.Println(len(msg.Name))
	fmt.Println(msg.Name[0].Url)
	if msg.Name[0].Url != url {
		t.Fatalf("data mismatch: %v != %v", msg, r2)
	}

	// Read Record r3
	if _, err := reader.ReadProtocolMessage(msg); err != nil {
		t.Fatal(err)
	}
	if msg.Docid != r3.Docid || msg.Payload != r3.Payload {
		t.Fatalf("data mismatch: %v != %v", msg, r3)
	}

	if _, err := reader.ReadProtocolMessage(msg); err != nil {
		fmt.Println("ok : here is a expected err ", err)
	}
	reader.Finish() // close underlying fd

}

File: cpp/recordio_test.cc
#include "recordio.h"

#include <string>
#include <boost/scoped_ptr.hpp>

#include "file.h"
#include "gtest/gtest.h"

#include "document.pb.h"

using namespace commonlib::util;

class RecordIOTest : public ::testing::Test {
protected:
    virtual void SetUp() {
        r1.set_docid(10);
        r1.mutable_links()->add_forward(20);
        r1.mutable_links()->add_forward(40);
        r1.mutable_links()->add_forward(60);
        r1.add_name();
        r1.mutable_name(0)->add_language();
        r1.mutable_name(0)->mutable_language(0)->set_code("en-us");
        r1.mutable_name(0)->mutable_language(0)->set_country("us");
        r1.mutable_name(0)->add_language();
        r1.mutable_name(0)->mutable_language(1)->set_code("en");
        r1.mutable_name(0)->set_url("http://A");
        r1.add_name();
        r1.mutable_name(1)->set_url("http://www.weibo.com/zhihu?from=feed&loc=nickname");
        r1.add_name();
        r1.mutable_name(2)->add_language();
        r1.mutable_name(2)->mutable_language(0)->set_code("en-gb");
        r1.mutable_name(2)->mutable_language(0)->set_country("gb");

        r2.set_docid(20);
        r2.mutable_links()->add_backward(10);
        r2.mutable_links()->add_backward(30);
        r2.mutable_links()->add_forward(80);
        r2.add_name();
        r2.mutable_name(0)->set_url("http://C");

        r3.set_docid(30);
        r3.mutable_links()->add_backward(100);
        r3.add_name();
        r3.mutable_name(0)->set_url("http://D");

        ::commonlib::util::file::ReadFileToString("./file.h", r3.mutable_payload());
    }
    virtual void TearDown() {
        m_reader.reset();
        m_writer.reset();
    }

protected:
    boost::scoped_ptr<RecordReader> m_reader;
    boost::scoped_ptr<RecordWriter> m_writer;
    Document r1;
    Document r2;
    Document r3;
};

TEST_F(RecordIOTest, TestNextMessageSequnce) {
    File *file = File::Open("./test.dat", "w+");
    ASSERT_TRUE(file->Open());

    m_writer.reset(new RecordWriter(file));

    ASSERT_TRUE(m_writer->WriteProtocolMessage(r1));
    ASSERT_TRUE(m_writer->WriteProtocolMessage(r2));
    ASSERT_TRUE(m_writer->WriteProtocolMessage(r3));

    ASSERT_TRUE(file->Close());
    delete file;

    LOG(INFO) << "END WRITE";
    
    file = File::Open("./test.dat",  "r");
    ASSERT_TRUE(file->Open());

    m_reader.reset(new RecordReader(file));

    Document message;
    ASSERT_TRUE(m_reader->ReadProtocolMessage(&message));
    LOG(INFO) << message.DebugString();
    ASSERT_TRUE(message.docid() == r1.docid());

    ASSERT_TRUE(m_reader->ReadProtocolMessage(&message));
    ASSERT_TRUE(message.docid() == r2.docid());

    ASSERT_TRUE(m_reader->ReadProtocolMessage(&message));
    ASSERT_TRUE(message.docid() == r3.docid());

    LOG(INFO) << message.DebugString();

    LOG(INFO) << "modifytime [" << file->ModifyTime();

    //ASSERT_TRUE(m_reader->ReadProtocolMessage(&message));
    ASSERT_TRUE(file->Close());
}

